### PR TITLE
Kovshov Homework

### PR DESCRIPTION
**Pull request by prohladenn**
Remove line if task was accepted earlier.

Mark task as done `[x]` if your resolve it.

- [ ] 1 ArrayDequeSimple
- [ ] 2 ArrayDequeSimple + Iterator Remove
- [ ] 3 ArrayDequeFull
- [x] 4 LinkedDequeSimple
- [x] 5 LinkedDequeSimple + Iterator Remove
- [x] 6 LinkedDequeFull
- [x] 7 ArrayPriorityQueueSimple
- [x] 8 ArrayPriorityQueueSimple + Iterator Remove
- [x] 9 MergingPeekingIncreasingIterator
- [x] 10 AVLTree
- [x] 11 AVLTreeIterable
- [x] 12 AVLTreeIterable + Iterator Remove
- [x] 13 RedBlackTree (add)
- [x] 14 RedBlackTree (add + remove)
- [x] 15 OpenHashTable + Student
- [x] 16 OpenHashTable + Student + Iterator Remove

### CODE REVIEW

File: src/ru/mail/polis/collections/set/sorted/todo/RedBlackTree.java [hunk 1]
```diff
@@ -44,11 +60,236 @@ public String toString() {
...
+            node.parent.left = leftChild;
+        } else {
+            node.parent.right = leftChild;
+        }
+
+        leftChild.right = node;
+        node.parent = leftChild;
+    }
+
+    private void adjustAfterInsertion(RBNode<E> node) {
```
Comment (mikhail-nechaev) [inline src/ru/mail/polis/collections/set/sorted/todo/RedBlackTree.java]: сколько тут отличий от — http://cs.lmu.edu/~ray/notes/redblacktrees/ ?

File: src/ru/mail/polis/collections/set/hash/todo/Student.java [hunk 2]
```diff
@@ -49,20 +49,34 @@ public int getGroupId() {
...
+        if (probId < 0 || probId >= tableSize) throw new IllegalArgumentException();
+        int hash = (hashCode() + probId * hashCode2()) % tableSize;
+        return hash < 0 ? hash + tableSize : hash;
     }
 
     @Override
     public int hashCode() {
-        //todo: don't forget [hashCode - equals] contract
-        throw new UnsupportedOperationException("todo: implement this");
+        int hash = 7;
```
Comment (mikhail-nechaev) [inline src/ru/mail/polis/collections/set/hash/todo/Student.java]: такое уже было. ну допустим

**Comment by mikhail-nechaev** [comment]
всё сданное, кроме
- [ ] 9 MergingPeekingIncreasingIterator
- [ ] 13 RedBlackTree (add)
- [ ] 14 RedBlackTree (add + remove)

**Comment by prohladenn** [comment]
Возможности до(пере)делать уже нет?

File: src/ru/mail/polis/collections/set/hash/todo/Student.java (hunk 2 above)
Comment (mikhail-nechaev) [reply src/ru/mail/polis/collections/set/hash/todo/Student.java]: а. нет. списано #5

**Comment by mikhail-nechaev** [comment]
списано 52f9726d544b1ac46bfc7c7f1712b1c49da826f0
- [ ] 10 AVLTree
- [ ] 11 AVLTreeIterable
- [ ] 12 AVLTreeIterable + Iterator Remove

**Pull request merged**